src/cpp/session/modules/SessionRmdNotebook.R

```diff
@@ -0,0 +1,186 @@
+#
+# SessionRmdNotebook.R
+#
+# Copyright (C) 2009-16 by RStudio, Inc.
+#
+# Unless you have received this program directly from RStudio pursuant
+# to the terms of a commercial license agreement with RStudio, then
+# this program is licensed to you under the terms of version 3 of the
+# GNU Affero General Public License. This program is distributed WITHOUT
+# ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+# MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+# AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+#
+#
+
+.rs.addJsonRpcHandler("extract_rmd_from_notebook", function(input, output)
+{
+   # if 'output' already exists, compare file write times to determine
+   # whether we really want to overwrite a pre-existing .Rmd
+   if (file.exists(output)) {
+      inputInfo  <- file.info(input)
+      outputInfo <- file.info(output)
+      
+      if (outputInfo$mtime > inputInfo$mtime)
+         stop("'", output, "' exists and is newer than '", input, "'")
+   }
+   
+   contents <- .rs.extractFromNotebook("rnb-document-source", input)
+   cat(contents, file = output, sep = "\n")
+   
+   .rs.scalar(TRUE)
+})
+
+.rs.addFunction("extractFromNotebook", function(tag, rnbPath)
+{
+   if (!file.exists(rnbPath))
+      stop("no file at path '", rnbPath, "'")
+   
+   contents <- readLines(rnbPath, warn = FALSE)
+   reDocument <- paste('<!--', tag, '(\\S+) -->')
+   idx <- grep(reDocument, contents, perl = TRUE)
+   if (!length(idx))
+      stop("no encoded content with tag '", tag, "' in '", rnbPath, "'")
+   
+   rmdEncoded <- sub(reDocument, "\\1", contents[idx])
+   caTools::base64decode(rmdEncoded, character())
+})
+
+.rs.addFunction("executeSingleChunk", function(options,
+                                               content,
+                                               libDir,
+                                               headerFile,
+                                               outputFile) 
+{
+  # create a temporary file stub to send to R Markdown
```

Comment:
My guess is that we are also going to want to specify some fig.width and fig.height options here that are either well suited to the current size of the notebook or large enough that they can be easily downscaled to a nice looking image no matter what the size of the notebook is.